workflow/src/main/java/bio/terra/tanagra/workflow/CountUtils.java

```diff
@@ -0,0 +1,160 @@
+package bio.terra.tanagra.workflow;
+
+import java.util.Iterator;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.Distinct;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.Sum;
+import org.apache.beam.sdk.transforms.join.CoGbkResult;
+import org.apache.beam.sdk.transforms.join.CoGroupByKey;
+import org.apache.beam.sdk.transforms.join.KeyedPCollectionTuple;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TypeDescriptors;
+
+public final class CountUtils {
+  private CountUtils() {}
+
+  /**
+   * Count the number of distinct auxiliary nodes that each primary node has.
```

Comment:
Renamed auxiliary -> occurrence and added an example here. I struggled with how much of the primary/occurrence/concept (e.g. person/condition_occurrence/condition) relationships should be documented here, because this indexing workflow is really just doing a `count` with a `group by` clause. We may want to use this logic for other entity relationships in the future (e.g. which days did we diagnose someone with covid = count condition_occurrences grouped by the date). So I tried to name things here in relation to what the workflow is doing, as opposed to what types of entities and relationships we're using this for now. I see this workflow as having 3 parts: the nodes we want to generate counts for (called them "primary"), the occurrences that we want to process (called them "occurrences"), and the field that we want to count distinct occurrences of (called this "what_to_count"). Suggestions for better names are much appreciated.

Reply:
This seems reasonable for now. Probably needs some broader naming discussion to avoid conflicts before moving to something else. A few observations:

1. Occurrence may need some first class treatment in the underlay. Dealing with full generality here at the indexing layer would otherwise require some very vague language or generic and maybe non-performant functionality, e.g. you might describe this as "a hierarchical summing entity aggregator which takes an arbitrary counter expression". Or expanded for this specific case (still maintaining a degree of performance): a hierarchical aggregator on an entity1 by way of a specification relationship1 to unique instances of entity2.
2. The primary target relationship seems important here; and this code could perhaps be simplified if we have a term for this - i.e. in OMOP, the person table. It's a third variable which perhaps could act as more of a global concept.
3. The relationship of concept <- occurrence -> person seems important, and might want some special treatment.

Reply:
I agree the person/concept/occurrence pattern needs first class treatment in the underlay. Tim and I are still discussing what this should look like, but the current requirement that users define essentially identical relationships for each version of this pattern (e.g. for conditions, procedures, devices, etc.) is something we want to do away with. We should have some concept of an occurrence group of entities, where the user just tells us which ones they are and we infer the relationships and indexing workflows needed.

I'm not sure yet how much of that logic should live in the workflow classes themselves, as opposed to the code that generates the list of workflow commands to run.

Also agreed on the need for some better names here to avoid conflicts. Maybe this should be part of the general underlay overhaul.